src/storage.rs

```diff
@@ -1,20 +1,41 @@
 //! Ring buffer implementation, that does immutable reads.
 
 use std::any::TypeId;
+use std::fmt;
 use std::ops::{Index, IndexMut};
 
 /// Ringbuffer errors
-#[derive(PartialEq, Debug)]
-pub enum RBError<T> {
+pub enum RBError<'a, T: 'a> {
     /// If a writer tries to write more data than the max size of the ringbuffer, in a single call
     TooLargeWrite,
     /// If a reader is more than the entire ringbuffer behind in reading, this will be returned.
     /// Contains the data that could be salvaged, and the amount of data that was lost.
-    LostData(Vec<T>, usize),
+    LostData(StorageIterator<'a, T>, usize),
     /// If attempting to use a reader for a different data type than the storage contains.
     InvalidReader,
 }
 
+impl<'a, T: 'a> fmt::Debug for RBError<'a, T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            RBError::TooLargeWrite => write!(f, "TooLargeWrite"),
+            RBError::InvalidReader => write!(f, "InvalidReader"),
+            RBError::LostData(..) => write!(f, "LostData"),
+        }
+    }
+}
+
+impl<'a, T: 'a> PartialEq for RBError<'a, T> {
+    fn eq(&self, other: &RBError<'a, T>) -> bool {
+        match (self, other) {
+            (&RBError::TooLargeWrite, &RBError::TooLargeWrite) => true,
+            (&RBError::InvalidReader, &RBError::InvalidReader) => true,
+            (&RBError::LostData(..), &RBError::LostData(..)) => true,
```

Comment:
Should they really be considered equal?
How do you use this?

Reply:
well, for some reason the compiler required PartialEq to be defined for me to be able to do assertions on the error type, I could do it by using match in all the tests but ..

Reply:
Well, you compare them in assertions :smile: so compiler has a good reason to require `PartialEq` to be implemented.
Maybe you should compare content of `LostData`?

Reply:
Otherwise you may compare only `Ok` results in your tests.
Have you considered using `collect<Result<Vec<_>, >()`?

Reply:
Well, that's where it gets tricky, because it has an Iterator for the data it could recover, and the Iterator has a reference to the storage buffer, so I'd have to have comparison implemented on the storage buffer itself, which is kind of weird imho.

Reply:
And for the most part, external users will use match for this, to capture the different error cases.

Reply:
What about using `collect<Result<Vec<_>, _>()` and never trying comparing errors? 😏

Reply:
I understand your gripe with it, I'm trying to figure out a good solution. I think I might just do match in my tests.